File: core/hough_line_transform.py
import numpy as np
import cv2
import time

def hough_line_transform(edges,voting_threshold=70):
    Y,X = np.where(edges != 0)
    # This changes their shape from (N,) to (N, 1)
    X = X[:, np.newaxis]
    Y = Y[:, np.newaxis]
    
    diagonal = round(np.sqrt((edges.shape[0])*(edges.shape[0]) + (edges.shape[1])*(edges.shape[1])))
    theta = np.arange(0,181,1)
    
    accumulator = np.zeros((diagonal*2+1, theta.shape[0]), dtype=int)
    
    radians = np.deg2rad(theta)
    cosines = np.cos(radians)
    sines = np.sin(radians)
    
    R = np.round(X * cosines + Y * sines).astype(int)
    R += diagonal
    R = R.ravel()
    
    theta_flat = np.tile(np.arange(181), len(X))
    np.add.at(accumulator, (R, theta_flat), 1)
    
    best_r,best_theta = np.where(accumulator>voting_threshold)
    
    return best_r-diagonal,best_theta

def draw_hough_lines(original_image, best_r, best_theta):
    output_image = cv2.cvtColor(original_image, cv2.COLOR_GRAY2BGR)
        
    for r, theta_deg in zip(best_r, best_theta):
        theta_rad = np.radians(theta_deg)
        a = np.cos(theta_rad)
        b = np.sin(theta_rad)
        
        x0 = a * r
        y0 = b * r
        
        # Stretch the lines across the image
        x1 = int(x0 + 2000 * (-b))
        y1 = int(y0 + 2000 * (a))
        x2 = int(x0 - 2000 * (-b))
        y2 = int(y0 - 2000 * (a))
        
        # Draw the line in RED (B=0, G=0, R=255)
        cv2.line(output_image, (x1, y1), (x2, y2), (0, 0, 255), 2)
        
    return output_image

File: core/hough_line_transform.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <pybind11/pybind11.h>
#include <pybind11/numpy.h>
#include <opencv2/opencv.hpp>
#include "../utils/hough_utils.hpp"

using namespace std;
using namespace cv;

pybind11::array_t<int> hough_line_transform(pybind11::array_t<uint8_t> data, int voting_threshold = 70) {
    
    Mat edges = numpy_to_cvmat(data);

    int rows = edges.rows;
    int cols = edges.cols;

    // Find the edges coord
    vector<Point> edge_points;
    for (int y = 0; y < rows; ++y) {
        // Using ptr for fast row access in OpenCV
        const uint8_t* row_ptr = edges.ptr<uint8_t>(y);
        for (int x = 0; x < cols; ++x) {
            if (row_ptr[x] != 0) {
                edge_points.push_back(Point(x, y));
            }
        }
    }

    int diagonal = round(sqrt(rows * rows + cols * cols));
    int num_rho = 2 * diagonal + 1;
    int num_theta = 181;

    vector<double> cosines(num_theta);
    vector<double> sines(num_theta);
    for (int t = 0; t < num_theta; ++t) {
        double radians = t * CV_PI / 180.0;
        cosines[t] = cos(radians);
        sines[t] = sin(radians);
    }

    vector<int> accumulator(num_rho * num_theta, 0);

    // Cast votes
    for (const auto& pt : edge_points) {
        for (int t = 0; t < num_theta; ++t) {
            int r = round(pt.x * cosines[t] + pt.y * sines[t]);
            r += diagonal;
            accumulator[r * num_theta + t]++;
        }
    }

    // Extract lines above threshold
    houghlines best_lines;
    for (int r_idx = 0; r_idx < num_rho; ++r_idx) {
        for (int t = 0; t < num_theta; ++t) {
            if (accumulator[r_idx * num_theta + t] > voting_threshold) {
                best_lines.r.push_back(r_idx - diagonal);
                best_lines.theta.push_back(t);
            }
        }
    }

    return houghlines_to_2d_numpy(best_lines);
}
